### DISCUSSION TITLE
unable to live update button data help

### DISCUSSION THREAD

**Discussion by xalteropsx** [Textualize/textual#1473] [Q&A]
https://drive.google.com/drive/folders/1eOEuNEU8DosBzfqYG8YmMjBtTH2YTUqD?usp=sharing

my project 3 file css,py and data their plz help how can i live update these data 
this is small snippet what i trying to make it live otherwise i need to keep reopen app to see changes :(
```    
def on_mount(self) -> None:
        self.update_datavb()

    def update_datavb(self) -> None:
        filename = 'data.dat'
        with open(filename, 'rb') as file:
            self.data2 = pickle.load(file)

            self.data_buttons = []

            for i in range(len(self.data2)):
                bnum = len(self.data_buttons) + 1
                karrot = self.data2[int(i)]
                karrovt = karrot[0]
                karrovt2 = karrot[1]
                x = karrovt.split("\\")[-1]
                button = Button(f"{i}|{x}|*{karrovt2}s", id=f"button-{bnum}")
                self.data_buttons.append(button)
                self.horiz_container.mount(button)

    def watch_datavb(self):
        self.update_datavb()

```

**Comment by davep** [comment]
Looking at your code it's not obvious to me what it is you're attempting to do specifically, and I'm struggling to follow some of the code (for example, `watch_datavb` doesn't seem to be related to anything and, if it was, it contains an infinite loop that's going to tie up your application).

Given that, I'm going to try and infer the more general question and see if I can address that.

It seems to me that you wish to be able to poll some external file and, when the content changes, update the display. As a simple start for this I think I'd be tempted to look at using [Textual's `set_interval`](https://textual.textualize.io/api/message_pump/#textual.message_pump.MessagePump.set_interval) to check in on the file every so often and update the display if something has changed. Or as an alternative I suspect I'd take a look at using something like [aiowatcher](https://pypi.org/project/aiowatcher/) to react to when the file is changed (I don't know that that will mix well with Textual -- but that's something I'd investigate if I were in your apparent position).

**Reply by xalteropsx** [reply]
no it not work like that  actually when i update the file and when i click those data id it show me new updated output not old 

the problem is on terminal ui which not rerender its name

**Reply by xalteropsx** [reply]
i tried to wrap it reactive(name[0]) but it fail  sorry i dont know much about terminal ui documentation also far behind to do anything :(

**Reply by davep** [reply]
Unfortunately then, from what you've said here, it's not really clear to me what it is you're asking.

**Reply by xalteropsx** [reply]
i just want to perform a reload on container or mount which has this loop whenever i perform update or delete  wait i make a small video so u can have clear understanding

**Reply by davep** [reply]
If you mean you want to be able to update the content of existing widgets within your DOM, with new data that you've loaded, you likely want to be looking at using [queries](https://textual.textualize.io/guide/queries/) to find back the widgets (or simply hold "pointers" to them as instance variables, etc) and update their content on the appropriate way ([`Button` can be updated via its `label`](https://textual.textualize.io/widgets/button/), for example, and so on).

Another route to take would be to [use reactivity](https://textual.textualize.io/guide/reactivity/).

**Reply by xalteropsx** [reply]
https://user-images.githubusercontent.com/103671642/210352216-48b7c52d-4e26-4eb9-8bc8-4e2a362ff408.mp4

**Reply by davep** [reply]
Sorry, but I'm really unclear as to what that video is trying to show.

**Reply by xalteropsx** [reply]
> If you mean you want to be able to update the content of existing widgets within your DOM, with new data that you've loaded, you likely want to be looking at using [queries](https://textual.textualize.io/guide/queries/) to find back the widgets (or simply hold "pointers" to them as instance variables, etc) and update their content on the appropriate way ([`Button` can be updated via its `label`](https://textual.textualize.io/widgets/button/), for example, and so on).
> 
> Another route to take would be to [use reactivity](https://textual.textualize.io/guide/reactivity/).

reactive wont work on  
```         
       karrovt2 = Reactive(karrot[1])
                x = karrovt.split("\\")[-1]
                button = Button(f"{i}|{x}|*{karrovt2}s", id=f"button-{bnum}")

**Reply by xalteropsx** [reply]
i tried to wrap whole data as reactive but didnt work lol sorry i dont have any clue if u can enlighten me i have refactor alot of code u can work with an ease to my demo

**Reply by xalteropsx** [reply]
also i tried the label as well didnt updated saying it a string not reactive

**Reply by davep** [reply]
I think you may need to have another careful read over the reactivity documentation as the above snippet doesn't really use it in the way it's designed. Reactive variables are declared on the class, but accessed and updated as instance variables. If you've not done so already I *highly* recommend working from start to finish [through the tutorial](https://textual.textualize.io/tutorial/) as one of the subjects it does cover is how to use reactivity.

**Reply by xalteropsx** [reply]
> I think you may need to have another careful read over the reactivity documentation as the above snippet doesn't really use it in the way it's designed. Reactive variables are declared on the class, but accessed and updated as instance variables. If you've not done so already I _highly_ recommend working from start to finish [through the tutorial](https://textual.textualize.io/tutorial/) as one of the subjects it does cover is how to use reactivity.

already did that tutorial twice and look 5-6 time to get any suggestion but no luck i really fed up with docs and example are so low to make something more reactivity application based on data

**Reply by davep** [reply]
I'm more than happy to help, but it is proving difficult to really understand what you're having a problem with. Do you think you could try and reduce what you're trying to do to the most basic example, so we can try and figure out where you're getting stuck? Or if that doesn't work for you, perhaps you could document your code here and explain what each part is for and what you're hoping it will do? For example, I'm seeing things in your code that are defined but never used, so it's not clear if you mean for them to be there and expect them to have some effect, or if they're leftover code from some earlier test, etc.

If you can provide a clean example of what you are attempting, document what the code is aiming to do, ideally relate it to the Textual documentation so I can know what's informed you, and explain what you expect to happen but what isn't happening; that would be super helpful.

Or... in the most basic way possible, try and explain what sort of application you're trying to build here, and perhaps I can provide a simple example of how you might go about it (something I have tried above, based on what you've told us so far).

**Reply by xalteropsx** [reply]
```
import pickle

from textual.app import App, ComposeResult
from textual.containers import Container, Horizontal, Vertical, Grid
from textual.reactive import reactive
from textual.widgets import Button, Header, Footer, Static, Input


class Error(Static):
    Error = reactive("result", layout=True)

    def render(self) -> str:
        return f"Success-, {self.Error}!"


class TeslaClass(Static):
    # def on_mount(self) -> None:
    #     self.update_datavb()
    #
    # def update_datavb(self) -> None:
    #     print('da')
    # def watch_datavb(self):
    #     self.update_datavb()
    #
    #     while True:
    #         self.update_datavb()

    #for updating need new data input
    def callbackx(self, data):
        self.teslav = data[0]
        self.timeleft = data[1]
        self.channel_all = data[2]
        return [self.teslav, self.timeleft, self.channel_all]

    def delete(self):
        filename = 'data.dat'
        with open(filename, 'rb') as file:
            data3 = pickle.load(file)
            dino = []
            if data3[int(self.input11.value)]:
                data3.pop(int(self.input11.value))
                dino.append(data3)
            with open(filename, 'wb+') as xd:
                for i in range(len(dino)):
                    pickle.dump(dino[i], xd)

    def updatex(self):
        filename = 'data.dat'
        with open(filename, 'rb') as file:
            data3 = pickle.load(file)
            dino = []
            Relife = self.callbackx(self.data)
            data3[int(self.input6.value)] = Relife
            dino.append(data3)

            with open(filename, 'wb+') as xd:
                for i in range(len(dino)):
                    pickle.dump(dino[i], xd)

    def on_button_pressed(self, event: Button.Pressed) -> None:
        # input 11 for deleting
        if event.button.id == "del":
            if self.input11.value != '':
                self.delete()
                self.query_one(Error).Error = 'deleted'

        # for updating input6
        if event.button.id == 'up':
            if self.input3.value != '' and self.input1.value != '' and self.input4.value != '' and self.input6.value != '':
                self.channel_all = [self.input4.value]
                self.data = [self.input1.value, self.input3.value, self.channel_all]
                self.updatex()
                self.input1.focus()
                self.horiz_container.refresh()

                # bnum = int(self.input6.value)
                # karrotpanda = self.input1.value
                # karrotsecond = self.input3.value
                # button = Button(f">{bnum}|{karrotpanda}|*{karrotsecond}s", id=f"button-{bnum}")
                # self.data_buttons.append(button)
                # self.button.unmounted()
                # self.horiz_container.mount(button)
                # data = self.data[int(self.input6.value)]
                # self.output.update(repr(data))
                self.query_one(Error).Error = self.input1.value
        else:

            self.query_one(Error).Error = 'yolo'

    def compose(self):
        self.data = []
        self.input1 = Input(placeholder="ur file:")
        self.input3 = Input(placeholder="ur time")
        self.input4 = Input(placeholder="ur data1")
        self.input6 = Input(placeholder="id to update:")
        self.input11 = Input(placeholder="id to delete:")
        self.update3 = Error()
        self.output = Static("")
        self.horiz_container = Horizontal(
            Button("Delete", id="del", variant="success"),
            Button("Update", id="up", variant="success"),
            id='horizontal'
        )

        filename = 'data.dat'
        with open(filename, 'rb') as file:
            self.data2 = pickle.load(file)

            self.data_buttons = []

            for i in range(len(self.data2)):
                bnum = len(self.data_buttons) + 1

                karrot = self.data2[int(i)]
                karrovt = karrot[0]
                karrovt2 = karrot[1]
                x = karrovt.split("\\")[-1]
                button = Button(label=f"{i}|{x}|*{karrovt2}s", id=f"button-{bnum}")
                self.data_buttons.append(button)
                self.horiz_container.mount(button)

        yield Container(

            Vertical(

                Container(self.horiz_container),
                id="left-pane",
            ),
            Horizontal(
                Container(self.input1,
                          self.input3,
                          self.input4,
                          self.update3,

                          ),
                id="top-right",
            ),

            Container(

                Grid(self.output),

                Container(self.input6, self.input11, id="bottom-right-final"
                          ),
                id="bottom-right",
            ),
            id="app-grid",
        )


class ModalApp(App):
    CSS_PATH = "Stealth.css"
    watch_css = True

    def compose(self) -> ComposeResult:
        yield Container(Header(), Container(TeslaClass()), Footer())


if __name__ == "__main__":
    app = ModalApp()
    app.run()

**Reply by xalteropsx** [reply]
make sure u have css and data.dat and use this  to check update and delete 
so to see changes 
``` 
python urscript.py

**Reply by davep** [reply]
Thanks. As I asked earlier: *"perhaps you could document your code here and explain what each part is for and what you're hoping it will do?"* -- it really would make it so much easier to help you out if you could explain clearly what it is you are trying to do with this code, and perhaps explain which parts of the code relate to that.

How about we try it like this? Give me some step-by-step instructions on what you want to be able to do, such that I can run your app and do those steps, follow your explanation of what you've done and why, and then explain the part you're unclear about?

I *think* you're saying you want an app that can load data, allow the user to modify it, save the data again, etc. That's very doable with Textual as the underlying framework (I've written a couple of apps myself that do that sort of thing -- [OIDIA](https://github.com/davep/oidia) and [Unbored](https://github.com/davep/unbored) for example), and in what I've written so far here I've attempted to address what I *think* are the aspects of this you're struggling with. You've made it clear that I've not comprehended you questions so well so I'd really appreciate it if you could try to be as clear and concise as possible about what particular thing you're stuck with.

**Reply by xalteropsx** [reply]
https://drive.google.com/drive/folders/1eOEuNEU8DosBzfqYG8YmMjBtTH2YTUqD?usp=sharing

# i have put also more commenting for easier understanding each function 
1st= download all the file here in 1 folder 
2nd = python arcv.py
3rd = follow these video

https://user-images.githubusercontent.com/103671642/210462340-8c235fab-9796-41c3-8de2-1b4ac6c8cf75.mp4

4rd = see it not live reloading need to reopen application to see changes
5th = hint fill 3 input before update // and for delete u dont need fill any data make them empty and just put id to delete

**Reply by xalteropsx** [reply]
tried to read OIDIA quiet similar but split of file make me confused if it was single file might be good way to learn

**Reply by davep** [reply]
From watching the video again, I still feel that what you need to be looking into some more is [performing queries on your UI and updating the widgets that are there](https://github.com/Textualize/textual/discussions/1473#discussioncomment-4580929), perhaps that in combination with reactivity. However, you said that that wasn't the case and it didn't work for you. So this is why I'm struggling to fully grasp what it is you're looking for help with.

As I read it a "live updating" interface that stores data to a file, etc, would generally build up a UI, load values from a file into that UI, update values and display them in the UI and store them in the background. For all the Textual code I've written, and I've seen others write, that's tended to involve doing DOM queries and updating the content of the UI elements, or in extreme cases performing a `remove` followed by a fresh `mount`.

Summary is: seeing what you've written here, I feel I've pointed you in the correct direction for what it is you want to do; on the other hand you seem to feel I've not addressed your query. In which case I think it's clear there's a communication issue here.

So... I think what I'm going to try and do, when I get a moment, is try and put together a very basic example of what I think it is you want to be doing -- just the basic concept -- and try and illustrate some of the concepts I feel you need to get to grips with.

**Reply by xalteropsx** [reply]
yes sir any day when u got time please illustrate any basic example add update and delete 

here follow by Rustybrooks
```
from textual.app import App
from textual.containers import Container, Horizontal, Vertical
from textual.widgets import Button, Header, Footer, Static, Input


class TeslaClass(Container):
    DEFAULT_CSS = """
      #horizontal {
          background: blue;
          width: auto;
          height: auto;
      }
    """

    def on_button_pressed(self, event: Button.Pressed) -> None:
        if event.button.id == "submit":
            bnum = len(self.data_buttons) + 1
            button = Button(f"Data {bnum}", id=f"button-{bnum}")
            self.data_buttons.append(button)
            self.horiz_container.mount(button)
            self.data.append([self.input1.value, self.input2.value, self.input3.value])
            self.input1.value = ""
            self.input2.value = ""
            self.input3.value = ""
            self.input1.focus()
        else:
            button_index = int(event.button.id.split('-')[1]) - 1
            data = self.data[button_index]
            self.output.update(repr(data))

    def compose(self):
        self.data = []
        self.input1 = Input(placeholder="drag ur file .exe or shortcut:")
        self.input2 = Input(placeholder="put ur cd location where npm script run:")
        self.input3 = Input(placeholder="how much time taken to open file type-in-second:")
        self.output = Static("")
        self.horiz_container = Horizontal(
            Button("Submit", id="submit", variant="success"),
            id='horizontal'
        )
        self.data_buttons = []

        yield Vertical(
            self.input1,
            self.input2,
            self.input3,
            self.output,
            self.horiz_container,
        )


class TeslaApp(App):
    def compose(self):
        yield TeslaClass()


if __name__ == "__main__":
    app = TeslaApp()
    app.run()

**Reply by davep** [reply]
> here follow by Rustybrooks

Sorry, I don't know what this means. Could you explain please?

**Reply by davep** [reply]
Okay, here's a *very* basic stand-alone example of the core ideas I'm mentioning above and wanting to be sure you're aware of. This very simple one-file application comprises of a collection of input fields, and a collection of labels. When you enter a value in one of the fields and hit the submit key (the <kbd>Enter</kbd> key in Textual) the input value will be updated in the corresponding `Label`, and the file on disk will also be updated.

When the application is loaded later, if you're in the same directory (for the sake of simplicity it just saves to the current working directory), it'll reload the data.

```python
from json    import dumps, loads
from pathlib import Path

from textual.app        import App, ComposeResult
from textual.containers import Vertical, Horizontal
from textual.widgets    import Header, Footer, Input, Label
from textual.events     import Blur

class SimpleLiveUpdate( App[ None ] ):
    """A very simple 'live update' example.

    This app creates a series if inputs, and a series of outputs which will
    simply reflect the input when the submit key (the Enter key) is pressed
    on an `Input` field. The data is also saved when this happens too. On
    reloading the application the data will be reloaded.
    """

    CSS = """
    Vertical {
        width: 1fr;
    }

    Label {
        border: round red;
        height: 3;
        width: 100%;
    }
    """

    TEST_VALUE_COUNT = 20
    """How many input/output fields to create."""

    TEST_DATA = Path( "test-data.json" )
    """The test data will be held in a file in the current directory."""

    def compose( self ) -> ComposeResult:
        yield Header()
        yield Horizontal(
            Vertical(
                *[ Input( id=f"in-{n}", placeholder="Enter data here" ) for n in range( self.TEST_VALUE_COUNT ) ]
            ),
            Vertical(
                *[ Label( id=f"out-{n}" ) for n in range( self.TEST_VALUE_COUNT ) ]
            )
        )
        yield Footer()

    def on_input_submitted( self, event: Input.Submitted ) -> None:
        # Extract out the number of input from its ID.
        _, in_num = input.id.split( "-" )
        # Find the corresponding label in the DOM and update it.
        self.query_one( f"#out-{in_num}", Label ).update( input.value )
        # Also write the data.
        self.TEST_DATA.write_text( dumps( [ input.value for input in self.query( Input ) ] ) )

    def on_mount( self ) -> None:
        # We're starting up, is there some test data?
        if self.TEST_DATA.exists():
            # There is! Lets load it and populate the interface.
            for num, value in enumerate( loads( self.TEST_DATA.read_text() ) ):
                self.query_one( f"#in-{num}", Input ).value = value
                self.query_one( f"#out-{num}", Label ).update( value )

if __name__ == "__main__":
    SimpleLiveUpdate().run()
```

Have a read/play with that and let me know what you make if it so we can be 100% sure we're talking about the same thing.

**Reply by xalteropsx** [reply]
thank u very much

**Reply by xalteropsx** [reply]
how can we delete ? now does it go same with update ? or i saw something like del word but no idea

**Reply by xalteropsx** [reply]
![image](https://user-images.githubusercontent.com/103671642/211132165-a5453fb2-3b1c-48bf-a903-b542fd2ae32e.png)
```     
mount:
                if self.input11.value != '':
                    for num, value in enumerate(data):
                        self.query_one(f"#out-{num}").remove()

button:
           in_num = int(self.input11.value)
                await self.query_one(f"#out-{in_num}").remove()
                self.refresh()
                self.horiz_container.refresh()

**Reply by xalteropsx** [reply]
this happen seems like when we delete id wont refresh how can we refresh  ?

**Reply by xalteropsx** [reply]
it happen if we try to delete same id twice

**Reply by xalteropsx** [reply]
```      
  if event.button.id == "del":
            if self.input11.value != '':
                self.delete()
                in_num = int(self.input11.value)
# this work  it work on same id delete multiple since it keep refresh id
                self.query_one(f"#out-{in_num - 1}").update(str(in_num))
#this not work it only delete once then crash after second same id attempt 
                await self.query_one(f"#out-{in_num}").remove()

**Reply by davep** [reply]
Unfortunately I'm struggling to properly understand your questions. However, looking at the error message you mention, and the code just above, it looks to me like you're doing a `remove` on a specific widget in your DOM, with a specific ID. That makes sense, if you want to `remove` that specific widget with that ID that's what you'd do. What I don't understand is why you'd then want to try and `remove` it again. The error message you're seeing, and what you're asking and the code you're posting seems to suggest you're trying to `remove` a widget with a given ID, then `remove` it again. I don't understand why you'd want to `remove` something you've already removed.

More generally: would I be correct in thinking that English isn't your first language? That's okay, that's not a problem. But I would ask that you perhaps allow for this as we work through this. What you're asking might not quite make sense to a native English speaker like myself. So if I seem to be asking very basic questions, allow for that: I'm trying really heard to properly understand what it is you're struggling with.

**Reply by xalteropsx** [reply]
```
 self.query_one(f"#out-{in_num - 1}").update(str(in_num))

actually this thing refresh the new id after deleting old id

like if i delete something i want my all id to be refresh in order to reperform .remove()

**Reply by xalteropsx** [reply]
if need an example i can give an small application and video which work
